Create new section on how to use HTTP in components (#336)
bytecodealliance/component-docs: component-model/src/language-support/using-http-in-components/rust.md

# Rust

## 1. Setup

Add the `wasm32-wasip2` target to the Rust toolchain.

```rust
rustup target add wasm32-wasip2
```

Install [`wasmtime`][wasmtime]. The Wasmtime CLI has a built-in HTTP server that supports serving WebAssembly HTTP components.

```console
curl https://wasmtime.dev/install.sh -sSf | bash
```

[wasmtime]: https://github.com/bytecodealliance/wasmtime#installation

## 2. Creating a Rust WebAssemly project

Create a new Rust project with `cargo new`:

```console
cargo new wasm-http-hello-world
cd wasm-http-hello-world
```

> Add [`wstd`][wstd], a Rust async standard library for Wasm Components.

Reviewer: Can you elaborate here what `wstd` does for us? Maybe explain that it prevents the need to directly pull the WASI HTTP WITs and directly use the bindings generated by wit-bindgen.

Author: wdyt?

<img width="796" height="384" alt="image" src="https://github.com/user-attachments/assets/30774a09-62b5-4d49-ad3c-6a3401defa4a" />

# Rust

## 1. Setup

Add the `wasm32-wasip2` target to the Rust toolchain.

```rust
rustup target add wasm32-wasip2
```

Install [`wasmtime`][wasmtime]. The Wasmtime CLI has a built-in HTTP server that supports serving WebAssembly HTTP components.

```console
curl https://wasmtime.dev/install.sh -sSf | bash
```

[wasmtime]: https://github.com/bytecodealliance/wasmtime#installation

## 2. Creating a Rust WebAssemly project

Create a new Rust project with `cargo new`:

```console
cargo new wasm-http-hello-world
cd wasm-http-hello-world
```

> Add [`wstd`][wstd], a Rust async standard library for Wasm components.

Reviewer 2: ```suggestion
Add [`wstd`][wstd], a Rust async standard library for Wasm components as a dependency with `cargo add`:
```

```console
cargo add wstd
```

> `wstd` provides bindings for `wasi:http` as an idiomatic Rust async API. That's why, in this guide, you won't add WIT files to your project nor depend on `wit-bindgen` directly.

Reviewer 2: ```suggestion
`wstd` provides idiomatic Rust bindings for WASI standard interfaces (`wasi:http`) to increase ease-of-use for Rust WebAssembly components. Since we are using `wstd`, we will not need to add WIT files or depend on [`wit-bindgen`](https://crates.io/crates/wit-bindgen) directly.
```

Slight re-wording of this paragraph

> [!NOTE]

> > You can build an HTTP component in Rust without `wstd`. You would have to define the interfaces of the component in the WIT language, fetch their dependencies with `wkg` and generate the Rust bindings with `wit-bindgen`. Both approaches are valid.

Reviewer: ```suggestion
> You can build an HTTP component in Rust without `wstd`. You would have to fetch the [WASI HTTP WIT files](https://github.com/WebAssembly/WASI/tree/main/proposals/http) and their dependencies with `wkg` and then generate the Rust bindings with `wit-bindgen`. Both approaches are valid.
```

> > You can build an HTTP component in Rust without `wstd`. You would have to define the interfaces of the component in the WIT language, fetch their dependencies with `wkg` and generate the Rust bindings with `wit-bindgen`. Both approaches are valid.

Reviewer 2: ```suggestion
> It is possible to build an HTTP component in Rust without `wstd`. Building a HTTP component without `wstd` would require defining the `wasi:http` imports/exports of the component in WIT, fetching  WIT dependencies with `wkg` and generating the Rust bindings with `wit-bindgen`. 
> 
> Both approaches are valid, but `wstd` offers superior developer experience, so we opt to use it here.
```

>
> You can also use both `wstd` and `wit-bindgen`. They are not mutually exclusive and can co-exist in the same project.

[wstd]: https://docs.rs/wstd/latest/wstd/index.html

## 3. Writing the HTTP handler

We will implement the HTTP handler in `src/main.rs`. The file should look like the following:
```rust
use wstd::http::{Body, Request, Response, Result, StatusCode};

// WASI HTTP server components don't use a traditional `main` function.
// They export a function named `handle` which takes a `Request`
// argument, and which may be called multiple times on the same
// instance. To let users write a familiar `fn main` in a file
// named src/main.rs, wstd provides this `wstd::http_server` macro, which
// transforms the user's `fn main` into the appropriate `handle` function.
#[wstd::http_server]
async fn main(req: Request<Body>) -> Result<Response<Body>> {

>     match req.uri().path_and_query().unwrap().as_str() {

Reviewer 2: ```suggestion
    match req.uri().path() {
```

We can use just `path()` here

        "/" => home(req).await,
        _ => not_found(req).await,
    }
}

async fn home(_req: Request<Body>) -> Result<Response<Body>> {
    Ok(Response::new("Hello, world!\n".into()))
}

async fn not_found(_req: Request<Body>) -> Result<Response<Body>> {
    Ok(Response::builder()
        .status(StatusCode::NOT_FOUND)
        .body(().into())
        .expect("builder succeeds"))
}
```

## 4. Compiling and running the component

Build the component:

```console
cargo build --release --target wasm32-wasip2
```

> The resulting `.wasm` file will be compiled in `target/wasm32-wasip2/release/wasm-http-hello-world.wasm`.

Reviewer 2: ```suggestion
The `.wasm` binary for the component can be found at `target/wasm32-wasip2/release/wasm-http-hello-world.wasm`.
```

> Run the following Wasmtime command, it will spin-up an HTTP server on [`http://0.0.0.0:8080`](http://0.0.0.0:8080/) that redirect requests to instances of our component.

Reviewer 2: ```suggestion
To run the component, we can use [`wasmtime`](https://github.com/bytecodealliance/wasmtime/), a reference implementation host that supports the Component Model.

In particular, we can use `wasmtime serve` subcommand, which will spin-up an HTTP server at `http://localhost:8080` which will use our component to fulfill web requests. `wasmtime` creates a *fresh* instance of the component every time a request is served.
```
